qcp/grovers_algorithm.py

```diff
@@ -0,0 +1,125 @@
+from qcp.matrices import DefaultMatrix
+import random
+import qcp.constants as c
+import qcp.tensor_product as tp
+import qcp.gates as g
+
+
+def pull_set_bits(n: int):
+    """
+    Creates a list of bits that would be set to 1
+    to make the number n
+    :param n: number
+    :return: list[int] list of bits
+    """
+    bits = []
+    count = 0
+    while n:
+        cond = n & 1
+        if cond:
+            bits.append(count)
+        n >>= 1
+        count += 1
+    return bits
+
+
+class Grovers:
+
+    def __init__(self, size, target_state):
+        """
+        This is an implementation of Grover's algorithm which efficiently
+        finds a specific item in a list of items. In this implementation
+        we use Grover's algorithm to increase the amplitude of the
+        "target_state" and reduce all others in a "size"-qubit system
+
+        :param size: number of qubits in our circuit
+        :param target_state: specific state we want to target/select
+        """
+
+        assert (target_state < (2 ** size))
+        self.size = size
+        self.target = target_state
+        self.state = self.initial_state()
+
+        self.oracle = self.single_target_oracle()
+        self.diffuser = self.diffusion()
+        self.max_reflections = self.size - 1
+        # can only reflect size-1 times to get maximum probability
+
+        self.circuit = self.construct_circuit()
+
+    def initial_state(self):
+        """
+        Creates a state vector corresponding to |0..0>
+        :return: returns state vector
+        """
+        entries = [[0] for _ in range(self.size)]
```

Comment:
Just realised I suggested a typo...
```suggestion
        entries = [[0] for _ in range(2 ** self.size)]
```

Reply:
Woops!